### PR TITLE
WPB-5695 Enforce group conversation permission for external partner role

### PR DESCRIPTION
**Pull request by battermann**
https://wearezeta.atlassian.net/browse/WPB-5695

Enforce permission external partner cannot rename a conversation.

In proteus, 1:1 conversations of team members are regular conversation under the hood. Therefore an external partner should be able to create a regular team conversation with exactly one other member (aka 1:1 conv). But they should never be able to add members to conversations, regardless of the protocol, or whether it's a team conversation or not. This will ensure they can't transform a 1:1 conversation with a deleted user into a group conversation where they are admin. This is not a problem for MLS because MLS has proper 1:1 conversations.

## Checklist

 - [x] Add a new entry in an appropriate subdirectory of `changelog.d`
 - [x] Read and follow the [PR guidelines](https://docs.wire.com/developer/developer/pr-guidelines.html)

### CODE REVIEW

**Review by mdimjasevic — COMMENTED** [review]
I read the description of what you did. Is there a specific role for an external partner within `wire-api` or a service (Brig or Galley)? It feels like there ought to be one. Then you wouldn't have to work around the existing roles. Instead, you'd be able to apply permission rules specific to this role. In particular, you wouldn't have to do the dance of temporarily promoting an external partner from the `wire_member` role to the `wire_admin` role, only to revert it once the member-adding operation is done.

**Review by mdimjasevic — COMMENTED** [review]
The names of the used data constructors starting with `DoNotUseDeprecated` immediately raise a flag: why rely on the deprecated things? Each such constructor has this as a comment:

```haskell
  -- NOTE: This gets now overruled by conv level checks  
```

If the names are just misleading, can you write a test that disproves them?

File: integration/test/Test/ExternalPartner.hs [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+testExternalPartnerPermissions = do
+  (owner, tid, u1 : u2 : u3 : _) <- createTeam OwnDomain 6
+
+  partner <- createTeamMemberWithRole owner tid "partner"
+
+  -- a partner should not be able to create conversation with more than 2 users
+  void $ postConversation partner (defProteus {team = Just tid, qualifiedUsers = [u1, u2]}) >>= getJSON 403
+
+  do
+    -- a partner can create a one to one conversation with a user from the same team
```
[the remark this answers is not in the document]
Comment (battermann) [reply integration/test/Test/ExternalPartner.hs:38]: Yes, exactly.

File: services/galley/src/Galley/API/Action.hs [hunk 2]
```diff
@@ -464,6 +468,10 @@ performAction tag origUser lconv action = do
 
       pure (mempty, action)
     SConversationRenameTag -> do
```
[the remark this answers is not in the document]
Comment (battermann) [reply services/galley/src/Galley/API/Action.hs:470]: Yes, and externals should not be allowed to change the name of a group conversation in any case. This should also be enforced by the backend.

**Comment by battermann** [comment]
> The names of the used data constructors starting with `DoNotUseDeprecated` immediately raise a flag: why rely on the deprecated things? Each such constructor has this as a comment:
> 
> ```haskell
>   -- NOTE: This gets now overruled by conv level checks  
> ```
> 
> If the names are just misleading, can you write a test that disproves them?

IMO the tests verify that the mechanism works.
I don't know why those permissions are flagged with that prefix. Maybe by mistake?
I would suggest removing the prefix.

File: services/galley/src/Galley/API/Action.hs (hunk 2 above)
[the remark this answers is not in the document]
Comment (mdimjasevic) [reply services/galley/src/Galley/API/Action.hs:470]: That renaming restriction sounds fine. Does that mean there should be checks for other conversation actions too?

[the remark this answers is not in the document]
Comment (battermann) [reply services/galley/src/Galley/API/Action.hs:470]: Renaming was part of the steps to reproduce the bug. That's why I added it here.

**Review by mdimjasevic — APPROVED** [review]
Ship it!

**Pull request merged**